src/flamegraph/color.rs

```diff
@@ -0,0 +1,272 @@
+    } else if s.contains("Perl") || s.contains(".pl") {
+        Palette::Green
+    } else if s.ends_with("_[k]") {
+        Palette::Orange
+    } else {
+        Palette::Red
+    }
+}
+
+fn handle_js_palette(s: &str) -> Palette {
+    if s.ends_with("_[j]") {
+        if s.contains("/") {
+            return Palette::Green
+        } else {
+            return Palette::Aqua
+        }
+    } else if s.contains("::") {
+        return Palette::Yellow
+    } else if let Some(ai) = s.find("/") {
+        if (&s[ai..]).contains(".js") {
+            return Palette::Green
+        }
+    }
+
+    if s.contains(":") {
+        Palette::Aqua
+    } else if s == " " {
+        Palette::Green
+    } else if s.contains("_[k]") {
+        Palette::Orange
+    } else {
+        Palette::Red
+    }
+}
+
+fn handle_wakeup_palette(_s: &str) -> Palette {
+    Palette::Aqua
+}
+
+fn handle_chain_palette(s: &str) -> Palette {
+    if s.contains("_[w]") {
+        Palette::Aqua
+    } else {
+        Palette::Blue
+    }
+}
+
+fn coefficients_for_palette(palette: &Palette, name: &str, v1: f32, v2: f32, v3: f32) ->
+    ((u8, u8, f32), (u8, u8, f32), (u8, u8, f32)) {
+    let real_palette = match palette {
+        Palette::Hot => return ((205, 50, v3), (0, 230, v1), (0, 55, v2)),
+        Palette::Mem => return ((0, 0, v3), (190, 50, v2), (0, 210, v1)),
+        Palette::Io => return ((80, 60, v1), (80, 60, v1), (190, 55, v2)),
+        Palette::Red => return ((200, 55, v1), (50, 80, v1), (50, 80, v1)),
+        Palette::Green => return ((50, 60, v1), (200, 55, v1), (50, 60, v1)),
+        Palette::Blue => return ((80, 60, v1), (80, 60, v1), (205, 60, v1)),
+        Palette::Yellow => return ((175, 55, v1), (175, 55, v1), (50, 20, v1)),
+        Palette::Purple => return ((190, 65, v1), (80, 60, v1), (190, 65, v1)),
+        Palette::Aqua => return ((50, 60, v1), (165, 55, v1), (165, 55, v1)),
+        Palette::Orange => return ((190, 65, v1), (90, 65, v1), (0, 0, v1)),
+        Palette::Java => handle_java_palette(name),
+        Palette::Perl => handle_perl_palette(name),
+        Palette::Js => handle_js_palette(name),
+        Palette::Wakeup => handle_wakeup_palette(name),
+        Palette::Chain => handle_chain_palette(name),
+    };
+
+    coefficients_for_palette(&real_palette, name, v1, v2, v3)
+}
+
+fn affine_transform(a: u8, b: u8, x: f32) -> u8 {
+    b + (a as f32 * x) as u8
+}
+
+fn color_from_palette(palette: &Palette, name: &str, v1: f32, v2: f32, v3: f32) -> String {
+    let ((r_b, r_a, r_x),
+         (g_b, g_a, g_x),
+         (b_b, b_a, b_x)) = coefficients_for_palette(palette, name, v1, v2, v3);
+
+    let red = affine_transform(r_a, r_b, r_x);
+    let green = affine_transform(g_a, g_b, g_x);
+    let blue = affine_transform(b_a, b_b, b_x);
+
+    format!("rgb({},{},{})", red, green, blue)
+}
+
+pub(super) fn color_map<'a>(palette: &Palette, hash: bool, name: &'a str, palette_map: &'a mut HashMap<String, String>) -> &'a str {
```

Comment:
Does the palette map _need_ to hold `String`?

Reply:
Also, can we make it hold the color directly instead of the string mapping of it (i.e., the `(r, g, b)`)

Reply:
Thinking a little more, I wonder if we want to wrap the palette map into its own struct. That way we could move all of the `::load` and `.save` logic into an `impl` for that struct too. Maybe even in its own file?

Reply:
If we store the RBG components, we need to recreate the `rgb(r, g, b)` String each time right?

Reply:
Not entirely sure what you mean here. We (currently) need to produce one `String` for every frame we produce, pretty much no matter what. While we may be able to fix that with a clever stack-allocated buffer (#33), I don't think storing the three `u8`s in the map instead is worse? It just changes when that allocation happens (I think?). It either happens when you load the palette, or when you print it. And I think it'll be easier for us to get rid of the `String` allocation when writing out the color.

Reply:
Currently, when using `color_map` one `String` is allocated for each function name: two frames with the same function will borrow the same `String`.

Reply:
I think I'd still prefer that we don't allocate a string at all, which I think we'll be able to do with #33 :)

Reply:
So we need to:
- parse the rgb string when loading the palette from a file to extract the RGB components `(u8, u8, u8)`
- when rendering, for each frame, retrieve (or compute) the RGB components and format them into a string
- when saving the palette, for each function name, format the associated RGB components into a string

Just making sure that this is the direction that you want to take.

Reply:
- Yes
 - Sort of -- given #37, for each frame we actually just have to write the number to the stack, and don't need to allocate. That's the key for why I'd like to see this change!
 - Yes

Reply:
Done ([here](https://github.com/jonhoo/inferno/pull/32/commits/ff9625bae8e6f1d291d026fea6653dbc09afb0ba), and [here](https://github.com/jonhoo/inferno/pull/32/commits/1c21eca0e3ca0cf308a6b80f97eef497d74f8d3a))